### PR TITLE
Feature/direction dependent block attachment

### LINKED ISSUE #430: Make it so direction dependent attached blocks only connect to the craft when facing the craft.
My main focus is on torches but this also applies to things such as tripwire hooks and ladders.

When a craft is formed it will join these types of blocks even when they aren't connected to the craft. This creates floating torches and floating vehicles.

It would be nice if during the craft formation these types of blocks were only joined to the craft if they were facing towards a block that is part of the craft and not use these blocks to attach further blocks.

### PR DESCRIPTION
**Pull request by Intybyte**
### Describe in detail what your pull request accomplishes
When detecting a craft it checks if it is a block like a ladder, lantern on an attachable block, if said blocks aren't attached to the craft, then don't add them to the craft.

### Related issues:
Fix #430

### Checklist
- [x] Tested

### CODE REVIEW

**Comment by DerToaster98** [comment]
Uhm how attached does it have to be? Cause if it is hardcoded to be directly attached that is hurtful to creative designs. Shipbuilders often use the debug stick to get blocks into states that aren’t possible in survival (e.g.: levers). Also is there a whitelist or tag where one can limit the blocks affected by this? If no you should add that to let builders keep their creative freedom

**Comment by TylerS1066** [comment]
This does work with ladders, but does not work with torches or signs, both of which I would like to see added.

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [inline Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:312]: Remove commented code from PR

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [inline Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:357]: try to avoid shorthand names, `bd` -> `blockData` for instance

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [inline Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:361]: Consider using pattern matched switch statements by increasing the java version to java 21 - it is a requirement for Minecraft 1.20+

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [inline Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:386]: Remvoe debug logging

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [inline Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:385]: Remove this if statement and inline the `continue` to every line you do `blockFacingCraft=false`, than remove `blockFacingCraft`

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [inline Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:358]: Add a `CraftType` entry to allowlist this process based on materials, and another one to blanket enable the feature.

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [inline Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:399]: Don't invert this conditional, it makes the review more complex

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [inline Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:428]: I'd like to see an `Optional<BlockFace> getSupportFace(BlockData data)` static method that exists somewhere outside of this class to simplify the implementation logic. The method should match against as many block types that require support as possible, and leave it to the user to configure which blocks are actually checked.

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (TylerS1066) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:361]: 1.18.2 only requires Java 17, so we can't use Java 21 features until we drop support for it.

Comment (oh-noey) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:361]: why not?

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (Intybyte) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:358]: Just to be clear i need to create 2 entries like this?
directionDependency: true
directionDependentMaterials:
 - ladder

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (Intybyte) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:420]: Should I put that into `MovecraftLocation`?

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java]: no, the bool one should just be something to make it so that all materials are tested by default

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java:420]: Nah, probably somewhere else

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (Intybyte) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java]: That would be insanity, I tried that and even if they are not facing a block inside a craft won't be put inside, the only way I could spot that was thanks to some robust logging that players don't have

**Comment by Intybyte** [comment]
> Uhm how attached does it have to be? Cause if it is hardcoded to be directly attached that is hurtful to creative designs. Shipbuilders often use the debug stick to get blocks into states that aren’t possible in survival (e.g.: levers). Also is there a whitelist or tag where one can limit the blocks affected by this? If no you should add that to let builders keep their creative freedom

1 block only, if you are worried stuff like this is still possible (i think that is what you are refering to)
immagine

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (Intybyte) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java]: I can add a TODO when ready to update to java 21

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/SupportUtils.java
Comment (oh-noey) [inline Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/SupportUtils.java:32]: Might as well do ```java
if (data instanceof FaceAttachable faceAttachable && data instanceof Directional directional) {``` for future-proofing

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (oh-noey) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java]: Sorry, I meant all reasonable materials - we can have a base list somewhere

Comment (Intybyte) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java]: I would say to only use the `directionDependentMaterials` option, and if it is not used, just used a small list of the vary basic stuff like ladder and wall signs.

**Comment by DerToaster98** [comment]
One smaller hint: You could check if the block is a instanceof Attachable, which also covers banners, ladders and so on, Also it extends Directional, so the cast and check is not necessary there. FaceAttachable is still needed for grindstone and levers.

Also check for Hangable which covers lanterns and mangrove sapplings and potentially other future blocks too.

https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/material/Attachable.html
https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/block/data/Hangable.html

Also i think you should include a per-craft blacklist to prevent blocks from getting this treatment in addition if you haven't added that already

**Comment by Intybyte** [comment]
> One smaller hint: You could check if the block is a instanceof Attachable, which also covers banners, ladders and so on, Also it extends Directional, so the cast and check is not necessary there. FaceAttachable is still needed for grindstone and levers.
> 
> Also check for Hangable which covers lanterns and mangrove sapplings and potentially other future blocks too.
> 
> https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/material/Attachable.html https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/block/data/Hangable.html
> 
> Also i think you should include a per-craft blacklist to prevent blocks from getting this treatment in addition if you haven't added that already
> 
> Edit: Saw your todo notes just rn, ignore the part about Hangable if Hangable isn't a thing in 1.18

It is better that we don't use MaterialData, it is very deprecated and not used anymore, tried using that in the past and the whole server lags just to reinitialize a whole section of ancient code, displays a warning like "If this is not an old plugin, this must be an error", basically starts speaking latin just to get your code to run.

The BlockData equivalent is only used by [HangingSign, Tripwire, TripwireHook], the last 2 aren't that important/used so making another cast just for one block, HangingSign, that isn't even used that much as normal signs are cheaper, I don't think it is worth the effort.

**Comment by DerToaster98** [comment]
> > One smaller hint: You could check if the block is a instanceof Attachable, which also covers banners, ladders and so on, Also it extends Directional, so the cast and check is not necessary there. FaceAttachable is still needed for grindstone and levers.
> > Also check for Hangable which covers lanterns and mangrove sapplings and potentially other future blocks too.
> > https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/material/Attachable.html https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/block/data/Hangable.html
> > Also i think you should include a per-craft blacklist to prevent blocks from getting this treatment in addition if you haven't added that already
> > Edit: Saw your todo notes just rn, ignore the part about Hangable if Hangable isn't a thing in 1.18
> 
> It is better that we don't use MaterialData, it is very deprecated and not used anymore, tried using that in the past and the whole server lags just to reinitialize a whole section of ancient code, displays a warning like "If this is not an old plugin, this must be an error", basically starts speaking latin just to get your code to run.
> 
> The BlockData equivalent is only used by [HangingSign, Tripwire, TripwireHook], the last 2 aren't that important/used so making another cast just for one block, HangingSign, that isn't even used that much as normal signs are cheaper, I don't think it is worth the effort.

Why so arrogant?
Anyway, Attachable isn't marked as deprecated as far as i can see, other classes in the same package are though. Yes, initializing that takes time, that's why if it is necessary you put that into the startup section of your plugin which gets called in a phase that takes long anyway, so it does not matter.
Arguing that blocks aren't imported based on a crafting recipe isn't really logical. Movecraft is also used a lot on creative servers where that stuff is going to be used potentially since it is a neat alternative to sign + any block for it to attach to.

Anyway, TL;DR: I don't think hard checking for a list of blocks is a good idea. Usually the blocks that are attached to anything share one thing in their code which in bukkit is likely some interface

**Comment by DerToaster98** [comment]
Anyway, i stated my thoughts on here cause i thought it was necessary, might also leave now that i did that

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (TylerS1066) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java]: > why not?
Paper (and Spigot) will not start up with Java 21 unless the server owner explicitly bypasses their check.  While I do so on our servers, I do not plan on asking all server owners running Movecraft to do so.

File: Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java
Comment (TylerS1066) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java]: Perhaps it would make sense to have `directionDependency` default to everything, but if specified will become a whitelist (or blacklist)?  That way we wouldn't need two separate flags.

Comment (Intybyte) [reply Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java]: I think that is what I am doing for now, there is a default whitelist with all reasonable materials (note: we can't do everything as I said because it would be insanity and the staff would be forced to override the flag and set it to panic as some crafts would just stop working, all directionals include furnaces chests and more which is problematic), and in case it is overriden player can specify a whitelist. 

Implementing a blacklist or using all materials here would create chaos in my opinion, imagine joining a movecraft server with this default to "everything" and you can't pilot any craft because you furnace/chest's opposite facing block is air

**Comment by CiaoCiaoBambina** [comment]
I write on behalf of those who consider themselves creative builders on creative servers.  To DerToaster's point, this change would not only be catastrophic, but the precedent it would set could very well be terminal for much of the population who opts for aesthetic designs integrated into Movecraft.  It must be acknowledged that not all users of Movecraft frequent survival servers where this is meaningless, but also that many of them, such as myself, did not even originate from the AP series.  For those of us with creative backgrounds foreign to this mentality, the aforementioned proposition would disrupt our process, and even turn many of us away.  The creative population may not represent the entire market subjected to Movecraft, but every portion counts. 
 I doubt any of you would wish to turn people away...

This implementation would already break many intricate designs grounded on the basis of aesthetics and not survival functionality.  If this were to be approved, what else would, in the future?  It is Pandora's box for the creative sphere.  Again, branching off from DerToaster, if debugging states were to be analysed, that would be another great blow against this niche.  Below, I have attached a couple brief images to elaborate.












This change, and further ones in line with this subjective mentality would ruin many — if not all — of these designs, and would turn away a good portion of the creative userbase that makes use of Movecraft, myself included.  I am not sure suppressing those that cast good and attractive light on the admittedly-convoluted community is a good thing.

So I regress; the ideal course of action, from my perspective, is not to merge this pull request with the original project, but, instead, to have it as a standalone module that can be added by server owners and administrators to the Movecraft plugin, creating an opt-in.
A second route would be to add a toggle to the module itself, should it be hardcoded in, but setting it up so that it is an opt-in and not an opt-out (ease of access/use).
The third option would be not to merge at all :).

I understand that my perspective is very niche, but DerToaster encouraged me to comment, so I stand here in defence of the creative community who choose aesthetics over survival functionality in a way that does not impact the other community in any capacity; that is all we ask of you, in return.

**Comment by oh-noey** [comment]
Please read the discussion before making a comment, the plan is to make this configurable.

**Comment by CiaoCiaoBambina** [comment]
> Please read the discussion before making a comment, the plan is to make this configurable as an option on craft type files.

Forgive me, but this was never explicitly mentioned other than in DerToaster's requests for it to be executed.  I see a lack of confirmation of this in any instance.

**Comment by oh-noey** [comment]
I, the person reviewing the PR as a maintainer of the project, requested it:
> Add a CraftType entry to allowlist this process based on materials, and another one to blanket enable the feature.
To be clear, I totally appreciate your dedication to ensuring we meet the needs of all players, and I appreciate how thorough your comment was. If you have any other things that currently restrict creative players, make an issue on it - while I can't guarantee someone would pick it up, having that kind of unique insight on currently unknown feature gaps would definitely be valuable.

**Comment by CiaoCiaoBambina** [comment]
Must have missed that, apologies.

File: api/src/main/java/net/countercraft/movecraft/craft/type/CraftType.java
Comment (TylerS1066) [inline api/src/main/java/net/countercraft/movecraft/craft/type/CraftType.java:401]: We use `Tags.SIGNS` elsewhere for this, I'd recommend that.

File: api/src/main/java/net/countercraft/movecraft/craft/type/CraftType.java
Comment (TylerS1066) [inline api/src/main/java/net/countercraft/movecraft/craft/type/CraftType.java:397]: Redstone and soul torches should be added by default.  I think it may be best to put the entire list into `Tags.java` just so we have our material sets defined together.

**Review by TylerS1066 — COMMENTED** [review]
Looks good to me, just have a few comments on the defaults and their location in the codebase.

**Review by TylerS1066 — APPROVED** [review]

**Pull request merged**
